### ISSUE TITLE
Tensor Core support for NVIDIA Volta architecture

### ISSUE THREAD

**Issue by egborbe** [tensorflow/tensorflow#15897]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: NO
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**:
5386775e64aac0bb5020974122645da900bc312a
- **Python version**: 3.5
- **Bazel version (if compiling from source)**:0.8.1
- **GCC/Compiler version (if comp6iling from source)**:5.4.0
- **CUDA/cuDNN version**:9.1 / 7.0.5
- **GPU model and memory**: Titan V
- **Exact command to reproduce**:

### Describe the problem
It is widely reported that using float16 on Nvidia Volta architecture comes only with x2 improvement instead of the expected x4 x8 improvement using Tensor Cores
https://github.com/tensorflow/benchmarks/issues/77
https://devblogs.nvidia.com/parallelforall/programming-tensor-cores-cuda-9/

I checked that Tensorflow master branch used 
cudnnGetConvolutionForwardAlgorithm
to get the best possible algorithm for the given GPU.
However I think either
cudnnGetConvolutionForwardAlgorithm_v7
or 
cudnnFindConvolutionForwardAlgorithmEx
should be used to fully utilize the Volta architecture.
Could you please check this issue with a Volta architecture GPU?

**Comment by shivaniag** [comment]
@reedwm do you have any comments on this?

**Comment by reedwm** [comment]
@yzhwang can you look into using cudnnGetConvolutionForwardAlgorithm_v7 or cudnnFindConvolutionForwardAlgorithmEx, and see why we don't get a 4x performance improvement on ResNet 50?

Also /CC @nluehr, can you comment?

**Comment by cy89** [comment]
@zheng-xq @poxvoculi is this request easy or near what you're doing already?

**Comment by zheng-xq** [comment]
The TensorFlow head is already including TensorCore support on Volta.

**Comment by yzhwang** [comment]
Hi @egborbe , as @zheng-xq said, we have already included TensorCore support on Volta in TensorFlow. Here I will explain more the difference between all the cudnn APIs to find/get internal convolution algorithms:
 - Any algorithm starts with cudnnGet will use heuristics given a specific input shape as well as the scratch memory limitation to provide an internal algorithm that performs the best. This is light-weighted but not accurate. E.g. We might ended up with an internal algorithm that doesn't perform the best.
 - Any algorithm starts with cudnnFind will actually run all the available internal algorithms and returns a list of algorithm names as well as their profiling results, sorted in terms of performance. This is accurate, but first of all, it is as slow as writing our own autotune framework, second, the metrics it uses is not flexible and behind the binary, so users cannot control it.

For the above reason, in TensorFlow, we use our own autotune framework, if the autotune is turned off, we use cudnnGet to find a heuristic:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/cuda/cuda_dnn.cc#L2303
If at cuda_dnn.cc level, an algorithm is set, that means one of the two things:
 1) We are running autotune and we need to profile this algorithm;
 2) We know this algorithm performs the best.

So current autotune framework has already considered using tensor op:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/cuda/cuda_dnn.cc#L2370
It will automatically being used with cudnn version >=7000. To disable it, user can set the envvar:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/cuda/cuda_dnn.cc#L559

The benefit of using our own autotune system is that we can setup arbitrary metrics, like the one we have right now (you can find details in the comments):
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/gpu_utils.h#L40

I hope this answers your question and I think we can close this issue.

**Comment by egborbe** [comment]
Hi ,

thank you for your attention.
I now understand the way TF handles Cuda cores
I did some further measurements and it turned out that NCHW - NHWC conversion is now consuming as much time as the convolution itself on Volta.
I tried with NCHW and the performance gain against Pascal GPU shot up immediatelly to 250% from 200%. However this is still far from what Nvidia reported with NVCaffe (370% improvement in inference).
EVen when I use NCHW order for convulotion I realized there were nchw kernel calls as seen from nvprof.
So I think there is still room for improvement.
Thanks and regards,
Gabor Bereczki

**Comment by egborbe** [comment]
Correction
Even when I use NCHW order for convolution I realized there were nhwc kernel calls as seen from nvprof.

**Comment by egborbe** [comment]
this is the nvprof output
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   34.28%  797.43ms       301  2.6493ms  2.6269ms  2.6636ms  volta_fp16_s884cudnn_fp16_256x64_ldg8_relu_f2f_exp_small_nhwc2nchw_tn_v1
                   30.73%  714.84ms       301  2.3749ms  2.3708ms  2.3818ms  volta_fp16_s884cudnn_fp16_128x128_ldg8_relu_f2f_exp_small_nhwc2nchw_tn_v1
                   21.63%  503.23ms      1204  417.96us  3.0400us  1.2488ms  void nchwToNhwcKernel<__half, __half, float, bool=1>(int, int, int, int, __half const *, __half*, float, float)
 

and this is the code I used
import tensorflow as tf
import time

df='NCHW'
def net(dt, filt):
    a=tf.random_uniform((16,1,256,256), dtype=dt)
    f=tf.random_uniform((filt,filt,1,64), dtype=dt)
    a=tf.nn.conv2d(a,f,[1,1,1,1],'SAME',data_format=df)
    f=tf.random_uniform((filt,filt,64,128), dtype=dt)
    f2=tf.random_uniform((filt,filt,128,64), dtype=dt)
    for i in range(100):
        a=tf.nn.conv2d(a,f,[1,1,1,1],padding='SAME',data_format=df)
        a=tf.nn.conv2d(a,f2,[1,1,1,1],padding='SAME',data_format=df)
    a=tf.reduce_sum(a)
    return a
f16=net(tf.float16,3)

s=tf.Session()
s.run(tf.global_variables_initializer())
#input("press key")
for i in range(3):
    s1=time.time()
    x=s.run(f16)
    e1=time.time()
    print("FP16",(e1-s1)*1000)

**Comment by yzhwang** [comment]
Thanks for the investigation. I will do some benchmark and give updates.

**Comment by zheng-xq** [comment]
Volta GPU actually has a format conversion by itself. It's from Cudnn, not TensorFlow. For certain shapes, it is actually faster to be in NHWC.

**Comment by yzhwang** [comment]
Currently TensorFlow assumes NCHW as the better data format to present tensors for convolution on the GPU. For Volta and later generation of GPUs, this may not be true any more. From my benchmark running on Volta machine with cudnn7, NHWC's performance is only shown to be better than NCHW for float32 with SAME padding, and the gap gets larger when number of channel gets larger.
We will work on adding autotune feature to choose between the better data format when doing convolution.

**Comment by Emma926** [comment]
Thanks @zheng-xq @egborbe @yzhwang for the helpful info!

Are Tensor cores are on by default using Tensorflow 1.7 or later? I run resnet50 with float16 and synthetic data on one node of v100 but only get 535 examples/sec, which is only half of what is reported [here](https://devblogs.nvidia.com/tensor-core-ai-performance-milestones/?linkId=100000002357715). In [this](https://github.com/tensorflow/benchmarks/issues/77) issue 675 examples/sec is reported. I wonder why people can get over 1k and whether 535 can be regarded as a reasonable number.
